[torch.distributed.rpc] Install method docstrings from PyRRef to RRef

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40461 [torch.distributed.rpc] Install method docstrings from PyRRef to RRef**

It turned out `:inheried-members:` (see [doc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass)) is not really usable. 

Because pybind11 generates a docstring that writes `self` as parent class, `rpc.PyRRef`, type.

As a workaround, I am pulling docstrings on parent-class, `PyRRef` class, into subclass, `RRef`. And do surgery on the docstring generated by pybind11.

![image](https://user-images.githubusercontent.com/7608630/85462788-bd8c6400-b55a-11ea-9295-c245194f3580.png)


Differential Revision: [D7933834](https://our.internmc.facebook.com/intern/diff/D7933834/)

xush6528 committed:
[torch.distributed.rpc] Install method docstrings from PyRRef to RRef

It turned out `:inheried-members:` (see [doc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass)) is not really usable.

Because it shows `self` as of parent type. See https://github.com/pytorch/pytorch/pull/35694#issuecomment-648320988.

I have submitted an issue to sphinx repo. https://github.com/sphinx-doc/sphinx/issues/7867.

As a workaround, I am pulling docstrings on parent-class, `PyRRef` class, into subclass, `RRef`.

Differential Revision: [D7933834](https://our.internmc.facebook.com/intern/diff/D7933834/)

[ghstack-poisoned]

xush6528 committed:
Update on "[torch.distributed.rpc] Install method docstrings from PyRRef to RRef"


It turned out `:inheried-members:` (see [doc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass)) is not really usable.

Because it shows `self` as of parent type. See https://github.com/pytorch/pytorch/pull/35694#issuecomment-648320988.

I have submitted an issue to sphinx repo. https://github.com/sphinx-doc/sphinx/issues/7867.

As a workaround, I am pulling docstrings on parent-class, `PyRRef` class, into subclass, `RRef`.

Differential Revision: [D7933834](https://our.internmc.facebook.com/intern/diff/D7933834/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -358,19 +359,49 @@ def _validate_rpc_args(backend, store, name, rank, world_size, rpc_backend_optio
-    class RRef(PyRRef, GenericWithOneTypeVar):
-        # Combine the implementation class and the type class.
-        pass
-except TypeError as exc:
```
Why don't we need the try-except any more?

xush6528 replied:
Because I am removing their inheritance relation.

xush6528 replied:
Have to not use inheritance, because when I call `super(RRef, self).method_name(*args, **kwargs)`, pybind11 expects an exact type match of `PyRRef`, while `self` is of `RRef` at that moment.

xush6528 committed:
Update on "[torch.distributed.rpc] Install method docstrings from PyRRef to RRef"


It turned out `:inheried-members:` (see [doc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass)) is not really usable.

Because it shows `self` as of parent type. See https://github.com/pytorch/pytorch/pull/35694#issuecomment-648320988.

I have submitted an issue to sphinx repo. https://github.com/sphinx-doc/sphinx/issues/7867.

As a workaround, I am pulling docstrings on parent-class, `PyRRef` class, into subclass, `RRef`.

![image](https://user-images.githubusercontent.com/7608630/85462788-bd8c6400-b55a-11ea-9295-c245194f3580.png)


Differential Revision: [D7933834](https://our.internmc.facebook.com/intern/diff/D7933834/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! Let's wait for tests.

xush6528 committed:
Update on "[torch.distributed.rpc] Install method docstrings from PyRRef to RRef"


It turned out `:inheried-members:` (see [doc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass)) is not really usable. 

Because pybind11 generates a docstring that writes `self` as parent class, `rpc.PyRRef`, type.

As a workaround, I am pulling docstrings on parent-class, `PyRRef` class, into subclass, `RRef`. And do surgery on the docstring generated by pybind11.

![image](https://user-images.githubusercontent.com/7608630/85462788-bd8c6400-b55a-11ea-9295-c245194f3580.png)


Differential Revision: [D7933834](https://our.internmc.facebook.com/intern/diff/D7933834/)

[ghstack-poisoned]

xush6528 committed:
Update on "[torch.distributed.rpc] Install method docstrings from PyRRef to RRef"


It turned out `:inheried-members:` (see [doc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoclass)) is not really usable. 

Because pybind11 generates a docstring that writes `self` as parent class, `rpc.PyRRef`, type.

As a workaround, I am pulling docstrings on parent-class, `PyRRef` class, into subclass, `RRef`. And do surgery on the docstring generated by pybind11.

![image](https://user-images.githubusercontent.com/7608630/85462788-bd8c6400-b55a-11ea-9295-c245194f3580.png)


Differential Revision: [D7933834](https://our.internmc.facebook.com/intern/diff/D7933834/)

[ghstack-poisoned]
